### PR TITLE
Gameplay

### PR DESCRIPTION
**Pull request by vlad2876**

### CODE REVIEW

File: src/app/components/home/game-container/skyline-area/ship/ship.component.ts [hunk 1]
```diff
@@ -10,15 +10,26 @@ export class ShipComponent implements OnInit {
...
-  @Input() shipType: ShipType = ShipType.BigShip;
+  @Input() ship: Ship;
 
   constructor() { }
 
   ngOnInit() {
   }
 }
+
+export class Ship {
```
Comment (Zmitserkos) [inline src/app/components/home/game-container/skyline-area/ship/ship.component.ts:26]: Here it's better to use an interface that contains all the necessary fields.

File: src/app/entities/GameData.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import {GameStatusType} from "../gameplay-enums/status-type.enum";
+import {SpeedType} from "../gameplay-enums/speed-type.enum";
+
+export class GameData {
```
Comment (Zmitserkos) [inline src/app/entities/GameData.ts:4]: Please use an interface instead.

File: src/app/entities/SeaBattleGame.ts [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+import {SeaBattleShot} from "./SeaBattleShot";
+import {GameData} from "./GameData";
+
+export class SeaBattleGame {
+
+  private ships: SeaBattleShip[] = [];
+  private shots: SeaBattleShot[] = [];
+  private sight: SeaBattleSight;
+
+  private points = 0;
```
Comment (Zmitserkos) [inline src/app/entities/SeaBattleGame.ts:17]: It's better to initialize the fields that are related to the game logic in the constructor.

File: src/app/entities/SeaBattleGame.ts [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  private status: GameStatusType;
+  private endDate = 0;
+  private gameDuration = 0;
+  private maxShotCount = 10;
```
Comment (Zmitserkos) [inline src/app/entities/SeaBattleGame.ts:21]: Please mark the field as readonly. Generally it's better to keep such constants in a separate class.

File: src/app/entities/SeaBattleGame.ts [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+  private shotCount = 0;
+  private shotRemainingCount = this.maxShotCount - this.shotCount;
+  private bigShipsDestroyed = 0;
```
Comment (Zmitserkos) [inline src/app/entities/SeaBattleGame.ts:24]: It's better to remove these 2 fields and make them dynamic (to declare in the interface for the statistics table).

File: src/app/entities/SeaBattleGame.ts [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    private username: string,
+    private maxGameTime: GameDuration,
+    private startDate: number,
+    private areaWidth: number,
+    private areaHeight: number,
+    private gameSpeed: SpeedType
+  ) {
+  }
+
+  getData(): GameData {
```
Comment (Zmitserkos) [inline src/app/entities/SeaBattleGame.ts:46]: Should return the object that implements the correspondent interface.

File: src/app/entities/SeaBattleGame.ts [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+      this.smallShipsDestroyed,
+      this.startDate,
+      this.endDate,
+      this.areaWidth,
+      this.areaHeight,
+      this.gameSpeed
+    );
+  }
+
+  countdownTimer() {
```
Comment (Zmitserkos) [inline src/app/entities/SeaBattleGame.ts:63]: **startTimer()** is more readable. Also make this method private.

File: src/app/entities/SeaBattleGame.ts [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+  }
+
+  makeShot() {
+    this.shotCount++;
+    this.shotRemainingCount = this.maxShotCount - this.shotCount;
+  }
+
+  startNewGame() {
```
Comment (Zmitserkos) [inline src/app/entities/SeaBattleGame.ts:71]: **startGame**() is enough.

File: src/app/entities/SeaBattleGame.ts [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+    this.runNewShip();
+  }
+
+  pauseGame() {
+    this.status = GameStatusType.Paused;
+  }
+
+  resumeGame(game: SeaBattleGame) {
```
Comment (Zmitserkos) [inline src/app/entities/SeaBattleGame.ts:79]: Please remove **game** parameter.

File: src/app/entities/SeaBattleGame.ts [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+  }
+
+  setSightPosition(leftIndent: number) {
+  }
+
+  setShipPosition(leftIndent: number) {
```
Comment (Zmitserkos) [inline src/app/entities/SeaBattleGame.ts:85]: Please add a parameter that identifies the ship.

File: src/app/entities/SeaBattleGame.ts [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+  }
+
+  setShotPosition(position: number) {
```
Comment (Zmitserkos) [inline src/app/entities/SeaBattleGame.ts:88]: Please add a parameter that identifies the shot.

File: src/app/entities/SeaBattleShip.ts [hunk 12]
```diff
@@ -0,0 +1,17 @@
+import {ShipType} from "../components/home/game-container/game-container-enums/ship-type.enum";
+
+export class SeaBattleShip {
+
+  private leftIndent: number;
```
Comment (Zmitserkos) [inline src/app/entities/SeaBattleShip.ts:5]: Please add a parameter that defines a ship movement direction.

File: src/app/entities/SeaBattleShip.ts [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+  private isDestroyed: boolean;
+  private type: ShipType;
+
+  constructor(leftIndent: number, isDestroyed: boolean, type: ShipType) {
+    this.leftIndent = leftIndent;
+    this.isDestroyed = isDestroyed;
+    this.type = type;
+  }
+
+  runNewShip() {
```
Comment (Zmitserkos) [inline src/app/entities/SeaBattleShip.ts:15]: Remove the method

File: src/app/entities/GameData.ts [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+
```
Comment (Zmitserkos) [inline src/app/entities/GameData.ts:5]: Please remove the empty line here and in other classes.

File: src/app/entities/SeaBattleGame.ts [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+  private nextShip = new Subject();
+  private score = new BehaviorSubject(this.points);
+  private shotRemaining = new BehaviorSubject(this.shotRemainingCount);
+  private timer = new BehaviorSubject(this.maxGameTime);
+  private selectedSpeed = new BehaviorSubject(this.gameSpeed);
+  private gameStatus = new Subject();
+  private shotAnimation = new Subject();
+  private shipAnimationState = new Subject();
+
+  constructor(
```
Comment (Zmitserkos) [inline src/app/entities/SeaBattleGame.ts:36]: Please use an object that implements the correspondent interface to initialize a class object.

File: src/app/entities/SeaBattleShip.ts [hunk 16]
```diff
@@ -0,0 +1,17 @@
+import {ShipType} from "../components/home/game-container/game-container-enums/ship-type.enum";
+
+export class SeaBattleShip {
+
+  private leftIndent: number;
+  private isDestroyed: boolean;
```
Comment (Zmitserkos) [inline src/app/entities/SeaBattleShip.ts:6]: It's better to create a new enum for ship 3 states.

File: src/app/components/home/game-container/skyline-area/ship/ship.component.ts (hunk 1 above)
Comment (vlad2876) [reply src/app/components/home/game-container/skyline-area/ship/ship.component.ts]: Fixed

File: src/app/entities/GameData.ts (hunk 2 above)
Comment (vlad2876) [reply src/app/entities/GameData.ts:5]: Fixed

File: src/app/entities/SeaBattleGame.ts (hunk 3 above)
Comment (vlad2876) [reply src/app/entities/SeaBattleGame.ts]: Fixed

File: src/app/entities/SeaBattleGame.ts (hunk 4 above)
Comment (vlad2876) [reply src/app/entities/SeaBattleGame.ts]: Fixed

File: src/app/entities/SeaBattleGame.ts (hunk 5 above)
Comment (vlad2876) [reply src/app/entities/SeaBattleGame.ts]: Fixed

File: src/app/entities/SeaBattleGame.ts (hunk 6 above)
Comment (vlad2876) [reply src/app/entities/SeaBattleGame.ts]: Fixed

File: src/app/entities/SeaBattleGame.ts (hunk 7 above)
Comment (vlad2876) [reply src/app/entities/SeaBattleGame.ts]: Fixed

File: src/app/entities/SeaBattleGame.ts (hunk 8 above)
Comment (vlad2876) [reply src/app/entities/SeaBattleGame.ts]: Fixed

File: src/app/entities/SeaBattleGame.ts (hunk 9 above)
Comment (vlad2876) [reply src/app/entities/SeaBattleGame.ts]: Fixed

File: src/app/entities/SeaBattleGame.ts (hunk 10 above)
Comment (vlad2876) [reply src/app/entities/SeaBattleGame.ts]: Fixed

File: src/app/entities/SeaBattleGame.ts (hunk 11 above)
Comment (vlad2876) [reply src/app/entities/SeaBattleGame.ts]: Fixed

File: src/app/entities/SeaBattleGame.ts [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+  setSightPosition(leftIndent: number) {
+  }
+
+  setShipPosition(leftIndent: number) {
+  }
+
+  setShotPosition(position: number) {
+  }
+
+  completeShot() {
```
[the remark this answers is not in the document]
Comment (vlad2876) [reply src/app/entities/SeaBattleGame.ts]: Fixed

File: src/app/entities/SeaBattleShip.ts (hunk 12 above)
Comment (vlad2876) [reply src/app/entities/SeaBattleShip.ts:6]: Fixed

File: src/app/entities/SeaBattleShip.ts (hunk 13 above)
Comment (vlad2876) [reply src/app/entities/SeaBattleShip.ts]: Fixed

File: src/app/entities/GameData.ts (hunk 14 above)
Comment (vlad2876) [reply src/app/entities/GameData.ts]: Fixed

File: src/app/entities/SeaBattleGame.ts (hunk 15 above)
Comment (vlad2876) [reply src/app/entities/SeaBattleGame.ts]: Fixed

File: src/app/entities/SeaBattleSight.ts [hunk 18]
```diff
@@ -0,0 +1,8 @@
+import {SightData} from "./SightData";
+
+export class SeaBattleSight {
+  constructor(private sightData: SightData) {
+    sightData.leftIndent = sightData.leftIndent ? sightData.leftIndent : 0;
+    sightData.isDisabled = sightData.isDisabled ? sightData.isDisabled : false;
```
Comment (Zmitserkos) [inline src/app/entities/SeaBattleSight.ts:6]: Boolean(sightData.isDisabled)

**Review by Zmitserkos — APPROVED** [review]

**Pull request merged**